### PR TITLE
CM-230: Add e2e test for operands override resources

### PR DESCRIPTION
**Pull request by lunarwhite**
Mainly refer to override args logic.
* add util funcs `addOverrideResources` and `verifyDeploymentResources `
* add 1 valid and 1 invalid cases for each operand: controller, cainjector, webhook

It's the initial attempt to add new e2e test in dev's repo.

Test profile: [aos-4_14/ipi-on-aws/versioned-installer-ci](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/ocp-common/job/Flexy-install/249313/)

<details>
<summary>Pass log</summary>

```
Overrides test
/Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:16
  When adding valid cert-manager controller override resources
  /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:146
    should add the resources to the cert-manager controller deployment
    /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:148
  > Enter [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:04:25.332
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:19 @ 11/30/23 00:04:25.332
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:23 @ 11/30/23 00:04:26.399
  < Exit [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:04:26.874 (1.54s)
  > Enter [It] should add the resources to the cert-manager controller deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:148 @ 11/30/23 00:04:26.874
  STEP: Adding cert-manager controller override resources to the cert-managaer operator object - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:150 @ 11/30/23 00:04:26.874
  STEP: Waiting for cert-manager controller status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:164 @ 11/30/23 00:04:27.359
  STEP: Waiting for the resources to be added to the cert-manager controller deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:168 @ 11/30/23 00:04:27.601
  < Exit [It] should add the resources to the cert-manager controller deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:148 @ 11/30/23 00:04:27.84 (965ms)
• [2.505 seconds]
------------------------------
Overrides test
/Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:16
  When adding valid cert-manager webhook override resources
  /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:174
    should add the resources to the cert-manager webhook deployment
    /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:176
  > Enter [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:04:27.84
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:19 @ 11/30/23 00:04:27.84
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:23 @ 11/30/23 00:04:28.918
  < Exit [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:04:31.667 (3.825s)
  > Enter [It] should add the resources to the cert-manager webhook deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:176 @ 11/30/23 00:04:31.667
  STEP: Adding cert-manager webhook override resources to the cert-managaer operator object - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:178 @ 11/30/23 00:04:31.668
  STEP: Waiting for cert-manager webhook controller status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:192 @ 11/30/23 00:04:32.468
  STEP: Waiting for the resources to be added to the cert-manager webhook deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:196 @ 11/30/23 00:04:44.045
  < Exit [It] should add the resources to the cert-manager webhook deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:176 @ 11/30/23 00:04:44.284 (12.612s)
• [16.437 seconds]
------------------------------
Overrides test
/Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:16
  When adding valid cert-manager cainjector override resources
  /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:202
    should add the resources to the cert-manager cainjector deployment
    /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:204
  > Enter [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:04:44.284
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:19 @ 11/30/23 00:04:44.284
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:23 @ 11/30/23 00:04:45.493
  < Exit [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:04:56.339 (12.053s)
  > Enter [It] should add the resources to the cert-manager cainjector deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:204 @ 11/30/23 00:04:56.339
  STEP: Adding cert-manager cainjector override resources to the cert-managaer operator object - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:206 @ 11/30/23 00:04:56.339
  STEP: Waiting for cert-manager cainjector controller status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:220 @ 11/30/23 00:04:56.817
  STEP: Waiting for the resources to be added to the cert-manager cainjector deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:224 @ 11/30/23 00:05:00.297
  < Exit [It] should add the resources to the cert-manager cainjector deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:204 @ 11/30/23 00:05:00.635 (4.296s)
• [16.349 seconds]
------------------------------
Overrides test
/Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:16
  When adding invalid cert-manager controller override resources
  /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:230
    should not add the resources to the cert-manager controller deployment
    /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:232
  > Enter [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:05:00.636
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:19 @ 11/30/23 00:05:00.636
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:23 @ 11/30/23 00:05:02.048
  < Exit [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:05:04.628 (3.992s)
  > Enter [It] should not add the resources to the cert-manager controller deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:232 @ 11/30/23 00:05:04.628
  STEP: Adding cert-manager controller override resources to the cert-managaer operator object - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:234 @ 11/30/23 00:05:04.628
  STEP: Waiting for cert-manager controller status to become degraded - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:246 @ 11/30/23 00:05:05.245
  STEP: Checking if the resources are not added to the cert-manager controller deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:250 @ 11/30/23 00:05:05.551
  < Exit [It] should not add the resources to the cert-manager controller deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:232 @ 11/30/23 00:05:05.799 (1.171s)
• [5.163 seconds]
------------------------------
Overrides test
/Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:16
  When adding invalid cert-manager webhook override resources
  /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:256
    should not add the resources to the cert-manager webhook deployment
    /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:258
  > Enter [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:05:05.799
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:19 @ 11/30/23 00:05:05.799
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:23 @ 11/30/23 00:05:06.82
  < Exit [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:05:07.395 (1.595s)
  > Enter [It] should not add the resources to the cert-manager webhook deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:258 @ 11/30/23 00:05:07.395
  STEP: Adding cert-manager webhook override resources to the cert-managaer operator object - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:260 @ 11/30/23 00:05:07.395
  STEP: Waiting for cert-manager webhook controller status to become degraded - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:272 @ 11/30/23 00:05:08.012
  STEP: Checking if the resources are not added to the cert-manager webhook deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:276 @ 11/30/23 00:05:08.251
  < Exit [It] should not add the resources to the cert-manager webhook deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:258 @ 11/30/23 00:05:08.492 (1.098s)
• [2.693 seconds]
------------------------------
Overrides test
/Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:16
  When adding invalid cert-manager cainjector override resources
  /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:282
    should not add the resources to the cert-manager cainjector deployment
    /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:284
  > Enter [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:05:08.493
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:19 @ 11/30/23 00:05:08.493
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:23 @ 11/30/23 00:05:09.464
  < Exit [BeforeEach] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:18 @ 11/30/23 00:05:09.954 (1.461s)
  > Enter [It] should not add the resources to the cert-manager cainjector deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:284 @ 11/30/23 00:05:09.954
  STEP: Adding cert-manager cainjector override resources to the cert-managaer operator object - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:286 @ 11/30/23 00:05:09.954
  STEP: Waiting for cert-manager cainjector controller status to become degraded - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:298 @ 11/30/23 00:05:10.442
  STEP: Checking if the resources are not added to the cert-manager cainjector deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:302 @ 11/30/23 00:05:10.681
  < Exit [It] should not add the resources to the cert-manager cainjector deployment - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:284 @ 11/30/23 00:05:10.978 (1.025s)
  > Enter [AfterAll] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:308 @ 11/30/23 00:05:10.979
  STEP: Reset cert-manager state - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:309 @ 11/30/23 00:05:10.979
  STEP: Waiting for operator status to become available - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:313 @ 11/30/23 00:05:12.076
  < Exit [AfterAll] Overrides test - /Users/yuewu/Documents/workspace/fork/cert-manager-operator/test/e2e/overrides_test.go:308 @ 11/30/23 00:05:12.727 (1.749s)
• [4.235 seconds]
------------------------------
```

</details>

### CODE REVIEW

File: test/e2e/utils_test.go
Comment (xingxingxia) [inline test/e2e/utils_test.go:208]:
```suggestion
// addOverrideResources adds the override resources to to the specific cert-manager operand. The update process
```

File: test/e2e/utils_test.go
Comment (xingxingxia) [inline test/e2e/utils_test.go:249]: 1. Your comment is "polls every 1 second" while argument is time.Second*5. I suggest 10 seconds.
2. Is it needed to check `IsNotFound` specially? Does the test have chance to see `IsNotFound`?

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:150]: Typo in `certmanagaer.operator` would be `certmanager.operator` instead.

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:178]: Ditto, typo.

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:206]: Same, typo :)

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:234]: Same, typo :)

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:260]: Another, same typo.

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:286]: Last one, typo.

File: test/e2e/utils_test.go
Comment (swghosh) [inline test/e2e/utils_test.go:235]: In the future this can cause a problem if we construct a complex test case (for some generic use case) that combines both `overrideArgs` and `overrideResources`. At that time we can check if `*Config` is nil / empty before proceeding to update with `&v1alpha1.DeploymentConfig{...}` directly.
But for now it should be fine, should block this PR (this is just me noting to myself :).

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:12]: We should add a newline here to separate the ginkgo/gomega imports from the other k8s and cert-manager imports.

Comment (swghosh) [reply test/e2e/overrides_test.go:12]: And, it would be good if k8s and our operator imports are in another block. Please refer: https://github.com/openshift/cert-manager-operator/blob/46069515b0bda3a5f51ba88705095a089258d07b/test/e2e/suite_test.go#L6-L24 for example.

File: test/e2e/overrides_test.go
Comment (swghosh) [inline test/e2e/overrides_test.go:307]: In future we can also add test cases for the following scenarios described in: https://github.com/openshift/cert-manager-operator/pull/136#discussion_r1281566526. Okay to skip for now as `TestMergeContainerResources` from https://github.com/openshift/cert-manager-operator/blob/master/pkg/controller/deployment/deploment_helper_test.go already tests similar functionality in a unit test.

**Review by swghosh — APPROVED** [review]
Thanks very much @lunarwhite for these additional test cases!
Minor nits, otherwise LGTM.

File: test/e2e/overrides_test.go
Comment (xingxingxia) [inline test/e2e/overrides_test.go:284]: Just a comment placeholder as tracker: as we talked offline, let's combine multiple `It`s into one single `It`, corresponding to the single test case describe in CM-230.

**Comment by swghosh** [comment]
/label qe-approved
(this PR from QE itself)
/label docs-approved
(no doc changes required by this PR)
/label px-approved
(no user facing changes introduced by this PR)

**Comment by swghosh** [comment]
@TrilokGeer need your approval to merge this PR.
TIA!

**Pull request merged**
